### DISCUSSION TITLE
support for offline playback

### DISCUSSION THREAD

**Discussion by emekoi** [librespot-org/librespot#610] [Feature Requests]
i have seen #152, but what i am asking is would it be possible to add support for playing the encrypted files that an official client has saved for offline playback? i.e. i tell the spotify client to store its data in `~/Spotify` and later on tell librespot to play audio using that `~/Spotify`? or would that go against the TOS and put librespot at risk? or if not, would using the same folder for 2 separate players, just clobber the data in the folder like the official client does?

**Comment by devgianlu** [comment]
I don't believe anyone has reversed how the files are encrypted and stored to disk by the official client.

**Comment by emekoi** [comment]
so the way that files are encrypted on disk is different from how they are sent over the network? or am i mistaken in thinking that Spotify simply sends encrypted audio data to clients, and they cache and decrypt that audio data and play it back? i thought that since the official client's cache and storage directory file structures were superficially similar to the ones librespot generates that they were the same.

that aside, do you think it would be possible for someone with little to no reverse engineering knowledge or experience (me) to reverse how the official client encrypts and stores files to disk? and if i did, would librespot merge such a PR?

**Comment by SuisChan** [comment]
Is this still interesting to anyone? I mean it would be nice to be able to get files from the official app cache and that would eliminate the need to store duplicate files and it is doable, but seriously what about the ToS?

**Reply by roderickvd** [reply]
We should not and will not pursue this beyond the caching functionality we already have of encrypted files - which requires an active account and network access to decrypt. Beyond our own interpretation of the ToS, Spotify has told us so.
